src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java

```diff
@@ -1,29 +1,75 @@
 package com.es.phoneshop.web;
 
-import com.es.phoneshop.model.product.ArrayListProductDao;
-import com.es.phoneshop.model.product.ProductDao;
+import com.es.phoneshop.cart.Cart;
+import com.es.phoneshop.cart.CartService;
+import com.es.phoneshop.cart.DefaultCartService;
+import com.es.phoneshop.dao.ArrayListProductDao;
+import com.es.phoneshop.dao.ProductDao;
+import com.es.phoneshop.exceptions.OutOfStockException;
 
 import javax.servlet.ServletConfig;
 import javax.servlet.ServletException;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import java.io.IOException;
+import java.text.NumberFormat;
+import java.text.ParseException;
 
 public class ProductDetailsPageServlet extends HttpServlet {
 
     private ProductDao productDao;
 
+    private CartService cartService;
+
     @Override
     public void init(ServletConfig config) throws ServletException {
         super.init(config);
         productDao = ArrayListProductDao.getInstance();
+        cartService = DefaultCartService.getInstance();
     }
 
     @Override
     protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
-        String pathInfo = request.getPathInfo();
-        request.setAttribute("product", productDao.getProduct(Long.parseLong(pathInfo.substring(1))));
+        Long id = parseProductId(request);
+        request.setAttribute("product", productDao.getProduct(id));
+        request.setAttribute("cart", cartService.getCart(request));
+        productDao.addToRecent(id, productDao.getRecentProductIds(request));
+        request.setAttribute("recentProducts", productDao.getRecentProducts(3, request));
```

Comment:
When I open product for the first time, it's already displayed as recently viewed. This can be confusing for user